Features/circular queue

SaxanLosterToreda opened the pull request.

SaxanLosterToreda committed:
A working circular queue with one logic test

SaxanLosterToreda committed:
added a size state and state validator

SaxanLosterToreda committed:
added more tests and an isInteger function

SaxanLosterToreda committed:
Finished Tests, changed isStateValid to isValidState and made it take an arg. small wording fix I noticed in priorityqueue tests

SaxanLosterToreda committed:
small changes to address some of the uncovered lines

SaxanLosterToreda committed:
added option to make push override front when cq is full

SaxanLosterToreda committed:
Updated changelog and readme

mbrich commented on src/circular-queue.ts:
```diff
@@ -53,8 +64,8 @@ export default class ArmorCircularQueue<T> implements ArmorCollection<T> {
-			if(!this.isValidState(result)){
-				throw new Error('options.state contains errors')
+			if (!this.isValidState(result)) {
+				throw new Error('options.state contains errors');
```
Provide specific error messages when throwing so the developer has a clear idea of what went wrong. Given you return null in most places `isValidState` is used (which is good), I recommend replacing the single throw inside this condition with the commented out block of throws below.

mbrich commented on tests/circular-queue.spec.ts:
```diff
@@ -0,0 +1,801 @@
+
+		it('should return false if state.size is not an integer >= 0', () => {
+			const custom = new ArmorCircularQueue<number>(maxSize);
+			const types = [-9, -1.5, -1, 0.5, 10.5, null, undefined];
```
Good coverage on the different types. 
Add a few extras:
strings: '0', '1', '-1, '-1.5', '1.5'
and the literal value NaN.
NaN in JS is technically a number, but usually the result of undefined operations or operations that can't be represented.

mbrich commented on tests/circular-queue.spec.ts:
```diff
@@ -0,0 +1,801 @@
+
+		it('should return false if state.maxSize is not an integer >= 1', () => {
+			const custom = new ArmorCircularQueue<number>(maxSize);
+			const types = [-9, -1.5, -1, 0, 0.5, 10.5, null, undefined];
```
Same additions to test types as above comment. Probably easier to add a constant at top of tests called 
```const INVALID_INT_VALUES = [];```
and use that for each test where testing values which are not integers.

mbrich commented on tests/circular-queue.spec.ts:
```diff
@@ -0,0 +1,801 @@
+
+		it('should return false if state.front is not an integer', () => {
+			const custom = new ArmorCircularQueue<number>(maxSize);
+			const types = [-1.5, 0.5, 7.5, null, undefined];
```
Same integer test comment applies here (see above).

mbrich commented on tests/circular-queue.spec.ts:
```diff
@@ -0,0 +1,801 @@
+
+		it('should return false if state.rear is not an integer', () => {
+			const custom = new ArmorCircularQueue<number>(maxSize);
+			const types = [-1.5, 0.5, 7.5, null, undefined];
```
Same integer test comment applies here (see above).

SaxanLosterToreda committed:
many small related changes to multiple files
set exports to default
updated collection interface
created clearElements and reset functions
moved interface functions towards bottom of files

SaxanLosterToreda committed:
fixed errors I forget to check for before last push
